rust/clients/bridge/src/relay_bot.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2021 Open Community Project Association https://ocpa.ch
+// This software is published under the AGPLv3 license.
+
+//! # Relay Bot functions
+//!
+//! Logging in and listening for the message on matrix room and sending messages from qaul.
+use super::rpc::Rpc;
+use matrix_sdk::{
+    room::Room,
+    ruma::events::{
+        room::message::{MessageEventContent, MessageType, TextMessageEventContent},
+        AnyMessageEventContent, SyncMessageEvent,
+    },
+    Client, ClientConfig, SyncSettings,
+};
+use prost::Message;
+use url::Url;
+/// include generated protobuf RPC rust definition file
+mod proto {
+    include!("../../../libqaul/src/rpc/protobuf_generated/rust/qaul.rpc.feed.rs");
+    include!("../../../libqaul/src/rpc/protobuf_generated/rust/qaul.rpc.users.rs");
+    include!("../../../libqaul/src/rpc/protobuf_generated/rust/qaul.rpc.rs");
+}
+
+async fn on_room_message(event: SyncMessageEvent<MessageEventContent>, room: Room) {
+    if let Room::Joined(room) = room {
+        let (msg_body, msg_sender) = if let SyncMessageEvent {
+            content:
+                MessageEventContent {
+                    msgtype: MessageType::Text(TextMessageEventContent { body: msg_body, .. }),
+                    ..
+                },
+            sender: msg_sender,
+            ..
+        } = event
+        {
+            (msg_body, msg_sender)
+        } else {
+            return;
+        };
+        if msg_sender != "@qaul-bot:matrix.org" {
+            let msg_text = format!("{} : {}", msg_sender, msg_body);
+            let proto_message = proto::Feed {
+                message: Some(proto::feed::Message::Send(proto::SendMessage {
+                    content: msg_text,
+                })),
+            };
+            // encode message
+            let mut buf = Vec::with_capacity(proto_message.encoded_len());
+            proto_message
+                .encode(&mut buf)
+                .expect("Vec<u8> provides capacity as needed");
+
+            // send message
+            Rpc::send_message(buf, super::rpc::proto::Modules::Feed.into(), "".to_string());
+
+            if msg_body.contains("!qaul") {
+                let content = AnyMessageEventContent::RoomMessage(MessageEventContent::text_plain(
+                    "I am a message sent from qaul network\n",
+                ));
+
+                // send our message to the matrix room we found the "!qaul command
+                room.send(content, None).await.unwrap();
+            }
+
+            if msg_body.contains("!users-list") {
+                // Here I need to trigger the RPC call from users.rs file.
```

Comment:
@MathJud This is the place where I am interested in calling the rpc function from `users.rs`. 
Use case is that. In matrix room, a user sends a message `!users-list`. This should return the data which we will format with better strings. 

If we are successful to do just this, Next we can call the rpc function of `feed.rs` and that should allow us to send any messages which we write in qaul world to be sent over to the matrix room. And this should be enough to complete our first milestone.

Reply:
There are several ways to do that.

1) The bridge client queries regularly libqaul via RPC for a list of users, e.g. all 10 seconds. This list is stored in a way to be accessed by different threads and functions (see the state storage object https://crates.io/crates/state )

2) Create an asynchronous event system, where you pause your function until the next user list arrived.

As we anyway will need the user information, and also will need to query several other information regularly I would go with solution 1. It is also much simpler to implement.

Reply:
Which users do you want to deliver here?

a) all known users, independent if they are online or not (this is an ever growing list of users)
b) the users that are online at this time.

Reply:
> Which users do you want to deliver here?
a) all known users, independent if they are online or not (this is an ever growing list of users)
b) the users that are online at this time.

It's (a)